src/test/kotlin/data/AuthenticationRepoImplTest.kt

```diff
@@ -0,0 +1,201 @@
+        }
+        assertEquals("User 'unknown' not found", exception.message)
+    }
+
+    @Test
+    fun `login should throw InvalidPasswordException when password is incorrect`() {
+        // Given
+        val username = "ahmed"
+        val correctHashed = "hashed-pass"
+        val wrongInput = "wrong-pass"
+        val storedUser = Mate(id = UUID.randomUUID(), username, correctHashed, Role.MATE)
+        every { usersDataSource.read() } returns listOf(storedUser)
+        every { hashingService.verifyData(wrongInput, correctHashed) } returns false
+
+
+        // When , Then
+        assertFailsWith<InvalidPasswordException> {
+            authRepo.login(username, wrongInput)
+        }
+    }
+
+    @Test
+    fun `createMate should fail if user already exists`() {
+        // Given
+        val admin = Admin(id = UUID.randomUUID(), "admin1", "hashed123", Role.ADMIN)
+        val existingUsername = "mate1"
+        val newMate = Mate(id = UUID.randomUUID(), existingUsername, "pass", Role.MATE)
+        every { usersDataSource.read() } returns listOf(newMate)
+
+        // When , Then
+        assertFailsWith<UserAlreadyExistException> {
+            authRepo.createMate(admin, existingUsername, "newPass")
+        }
+    }
+
+    @Test
+    fun `createMate should hash password and return correct Mate`() {
+        // Given
+        val admin = Admin(UUID.randomUUID(), "admin1", "hashed123", Role.ADMIN)
+        val username = "newMate"
+        val rawPassword = "password"
+        val expectedHash = "hashedPassword"
+        every { usersDataSource.read() } returns emptyList()
+        every { hashingService.hashData(rawPassword) } returns expectedHash
+        every { usersDataSource.write(any<Mate>()) } returns Unit
+
+        // When
+        val mate = authRepo.createMate(admin, username, rawPassword)
+
+        // Then
+        assertEquals(username, mate.username)
+        assertEquals(expectedHash, mate.password)
+        verify(exactly = 1) { hashingService.hashData(rawPassword) }
+    }
+
+    @Test
+    fun `createMate should add mate to users list`() {
+        // Given
+        val admin = Admin(UUID.randomUUID(), "admin1", "hashed123", Role.ADMIN)
+        val username = "newMate"
+        val rawPassword = "password"
+        val expectedHash = "hashedPassword"
+
+        val capturedMate = slot<Mate>()
+        var called = false
+
+        every { usersDataSource.read() } answers {
+            if (!called) {
+                called = true
+                emptyList()
+            } else {
+                listOf(capturedMate.captured)
+            }
+        }
+        every { hashingService.hashData(rawPassword) } returns expectedHash
+        every { usersDataSource.write(capture(capturedMate)) } returns Unit
+
+        // When
+        val mate = authRepo.createMate(admin, username, rawPassword)
+
+        // Then
+        assertTrue(usersDataSource.read().contains(mate))
+    }
```

Comment:
This can be ignored for now with `@Ignore` as it's considered an integration test.